### PR TITLE
FEATURE: Keep tags and collections of assets when exporting sites

### LINKED ISSUE #1097: Export media asset tags together with assets
Jira issue originally created by user niko5:

When exporting a Site tags are not exported for assets, this also affects the Import service as the are not imported.

 Jira-URL: https://jira.neos.io/browse/NEOS-220

### PR DESCRIPTION
**Pull request by kdambekalns**
With this change tags and collections of assets are included in site exports to XML and are imported properly, too.

See also #2503 and #2537.
Fixes #1097

### CODE REVIEW

**Comment by kdambekalns** [comment]
This replaces #2537 with a version that applies cleanly to 4.3. These are the original PR notes@zickgraf left:

---

**How I did it**

* Added converters from tags and asset collections to arrays:
  * For tags, the child property `assetCollections` is ignored because this would lead to an infinite loop.
  * For collections, the child property `assets` is ignored because this would lead to an infinite loop.
* Added a converter from array to tag. See the comment in https://github.com/neos/neos-development-collection/commit/3c5a4bb0d09bea16bcfdeceea6dc397adae662a9#diff-5d70864d51a35d6ab9fbb90ae189da11R57 for the reason why this is needed.
* For asset collections the generic converter from arrays suffices.

**How to verify it**

1. Add some tags and collections and assign some assets to them.
2. Export to `Sites.xml`.
3. Do a clean import.
4. Check that the tags and collections are preserved.

**Notes**
1. Of course, tags and collections not accessible from content are lost because the assets are lost.
2. I tried to test as many edge cases as possible but I'm not sure I have caught all of them. However, a clean import should work in any case so I don't expect any data loss even if some edge cases are not handled properly.

**Comment by kitsunet** [comment]
Seems a bit risky change for a bugfix release? I will have a close look ASAP...

**Comment by kdambekalns** [comment]
> Seems a bit risky change for a bugfix release? I will have a close look ASAP...

Well, might be. It has been rotting away for some while and I had it on my review-list, so I tackled it today…

**Comment by zickgraf** [comment]
Thanks for rebasing this! :-)

**Comment by macebu** [comment]
Two months later, can anyone find time for a second review?

File: Neos.Media/Classes/TypeConverter/TagToArrayConverter.php [hunk 1]
```diff
@@ -14,6 +14,7 @@
 use Neos\Flow\Annotations as Flow;
 use Neos\Flow\Persistence\PersistenceManagerInterface;
 use Neos\Flow\Property\PropertyMappingConfigurationInterface;
+use Neos\Flow\Property\TypeConverter\AbstractTypeConverter;
 use Neos\Media\Domain\Model\Tag;
 use Neos\Flow\Property\TypeConverter\AbstractTypeConverter;
```
Comment (zickgraf) [inline Neos.Media/Classes/TypeConverter/TagToArrayConverter.php]:
```suggestion
```

This was moved to line 17.

File: Neos.Media/Classes/TypeConverter/AssetCollectionToArrayConverter.php [hunk 2]
```diff
@@ -14,6 +14,7 @@
 use Neos\Flow\Annotations as Flow;
 use Neos\Flow\Persistence\PersistenceManagerInterface;
 use Neos\Flow\Property\PropertyMappingConfigurationInterface;
+use Neos\Flow\Property\TypeConverter\AbstractTypeConverter;
 use Neos\Media\Domain\Model\AssetCollection;
 use Neos\Flow\Property\TypeConverter\AbstractTypeConverter;
```
Comment (zickgraf) [inline Neos.Media/Classes/TypeConverter/AssetCollectionToArrayConverter.php]:
```suggestion
```

This was moved to line 17.

**Comment by macebu** [comment]
> 
> 
> I have suggested two changes which hopefully fix the CI errors.

Thanks @zickgraf  - /cc @kdambekalns and @kitsunet could you please review this?

**Comment by kdambekalns** [comment]
I cannot review, because technically this is "my" PR, from GH's perspective… 🤷‍♂️
But I'll check those build errors now.

**Comment by daniellienert** [comment]
Hey guys, 
personally, I don't think this is a bugfix, but adding a missing feature. As @kitsunet already stated, with that additional code and changing doctrine beahviour it is also very risky for a bugfix.  I would recommend to rebase it to master and release that feature with Neos 5.2

**Comment by zickgraf** [comment]
The commits apply cleanly to master (see https://github.com/zickgraf/neos-development-collection/tree/export-tags-collections) and I have run a successful test! Should I open a new PR or will we adjust this one?

A note regarding the current CI error: it looks like an intermittent error to me and I guess the CI will be green again after the next run.

**Comment by daniellienert** [comment]
Restarted travis

**Comment by kdambekalns** [comment]
That's what I did yesterday… Travis needs to pick up the changed branch, it doesn't do that when restarting. Let's try turning it on and off again… ;)

**Comment by kdambekalns** [comment]
Screenshot 2020-02-12 at 09 28 44

**Comment by zickgraf** [comment]
Thanks @daniellienert for the review and @kdambekalns for the merge :-)
